assignments/week_3/project_3_4/README.md

```diff
@@ -0,0 +1,102 @@
+# Project 3.4
+Welcome to project 3! In this week's project we will be going into graph search algorithms (breadth-first and depth-first search). These algoritms are an important graph search algorithm that can be used to solve many problems including finding the shortest path in a graph and solving puzzle games (such as Rubik's Cubes). Many problems in computer science can be thought of in terms of graphs. For example, analyzing networks, mapping routes, and scheduling are graph problems.
+
+# Introduction To Graph Search Algorithms (Optional Read)
+## Graph Data Structure
+Graphs are mathematical structures that represent pairwise relationships between objects. A graph is a flow structure that represents the relationship between various objects. Before we proceed further, let's familiarize ourselves with some important terms:
+- Vertex/Node − Each node of the graph is represented as a vertex. In the following example, the labeled circle represents vertices. Thus, A to G are vertices. We can represent them using an array as shown in the following image. Here A can be identified by index 0. B can be identified using index 1 and so on.
+- Edge − Edge represents a path between two vertices or a line between two vertices. In the following example, the lines from A to B, B to C, and so on represents edges. We can use a two-dimensional array to represent an array as shown in the following image. Here AB can be represented as 1 at row 0, column 1, BC as 1 at row 1, column 2 and so on, keeping other combinations as 0.
+- Adjacency − Two node or vertices are adjacent if they are connected to each other through an edge. In the following example, B is adjacent to A, C is adjacent to B, and so on.
+- Path − Path represents a sequence of edges between the two vertices. In the following example, ABCD represents a path from A to D.
+
+![Graph](images/graph.jpg)
+
+## Breadth-First Search (BFS)
+Breadth-first search (BFS) is an algorithm for traversing or searching tree or graph data structures. It starts at an initial node and explores all of the neighbor nodes at the present depth prior to moving on to the nodes at the next depth level. Below is an example that shows the order in which the nodes would be expanded into within BFS:
+
+![BFS Order](images/bfs_order.png)
+
+A standard BFS implementation puts each vertex of the graph into one of two categories: visited and not visited. The purpose of the algorithm is to mark each vertex as visited while avoiding cycles. The algorithm works as follows:
```

Comment:
to avoid cycles? Then I would mention why we care about avoiding cycles and what they are.